crates/extensions/c8y_mapper_ext/src/operations/topic_filter.rs

```diff
@@ -0,0 +1,43 @@
+//! Operations' topic filters.
```

Comment:
Not quite happy with this module with all the operation specific filters. I see that you wanted to separate the conversion aspects from the command handling aspects. In that case I get a feeling that we need two separate operation specific modules for each operation/command type: one with just the conversion aspects and other with the command processing aspects as done now.

Reply:
Yeah, I wanted to separate these because reacting to operation status changes and MQTT topics are really different concerns, and I went on a bit of an autopilot separating modules, but looking at these functions, they're completely superfluous, the implementation is the same, just with different arguments.

> I get a feeling that we need two separate operation specific modules for each operation/command type: one with just the conversion aspects and other with the command processing aspects as done now.

I think that wrt. the conversion aspect, there are really two things here: the topic filters (`topic_filter` module) and the conversions of incoming cumulocity messages that initiate thin-edge.io operations (`convert` module).

## On the topic fiilters:

I'm not sure if we need a conversion module per operation type. Topic filters are necessary to create an MQTT subscription, and in this case I think the pattern is more like:
- `OperationHandler` handles some set of operations
- if `CumulocityConverter` gets an operation of type that can be handled by `OperationHandler`, it should pass the message to the handler
- so `OperationHandler` needs to expose the set of operations it supports, so that `CumulocityConverter` makes a proper MQTT subscription
- but it doesn't have to be per operation type. I think what's preferable is that `OperationHandler` produces only one topic filter, that covers all the operations that are enabled by the current configuration.

Also this will hopefully clear up more once we've made `OperationHandler` an actor: the topic filters will fit neatly into the message types of the new operations actor.

## On the conversion functions in the `convert` module:

These are impl blocks on `CumulocityConverter`, which I've pulled under `operations` module because other than converting between Cumulocity messages and thin-edge.io messages, they perform some side effects like querying the entity store or calling into c8y proxy to obtain a URL where the operation has to upload a file.

This _auxiliary communication with C8y when we receive an operation_ is very much symmetric to _auxiliary communication with C8y when we terminate the operation (support or failure)_, so the long term plan is definitely to move these side-effects into `OperationHandler`, but right now it's very much coupled to `CumulocityConverter` which uses it primarily to convert between MQTT messages, so I think refactoring this will be best left as a follow-up.

## Conclusion

- as for the topic filters, instead of having different topic filters for different operation types, there should be `OperationHandler::topic_filter` that just creates a filter for all operation types the handler is interested in, and `CumulocityConverter` should just call it and don't care about different operation types.
- as for the `convert` module functions, I think this should be left as a follow-up.

How does that sound?

Reply:
`implementation` renamed in 80e6962, `OperationHandler::topic_filter` introduced in 6c7d4ae